en/_posts/2024-05-23-understanding-github-artifact-attestations.md

```diff
@@ -0,0 +1,224 @@
+for script injection attacks. While they can’t be used to attack the provenance
+itself, they could be used to attack the build without leaving any trace in the
+provenance.
+
+This means that while GitHub Artifact Attestations can be used with policy
+engines/tools like Rego, Cue, and OPA, these policies can’t be evaluated on the
+build inputs. For example, to verify that the build had no inputs etc. in order
+to prevent these kinds of attacks.
+
+This architecture also relies heavily on Sigstore’s Fulcio Certificate
+Authority to map the OID claims. This means that it’s hard to use separate
+private PKI, such as static keys, to sign and verify provenance. While this
+isn’t likely a problem for Open Source projects, some enterprises have these
+kinds of requirements.
+
+## Achieving SLSA Build L3 with GitHub Artifact Attestations
+
+In the documentation [SLSA Build L3 is described as
+achievable](https://github.blog/2024-05-02-introducing-artifact-attestations-now-in-public-beta/#an-effortless-user-experience)
+but it focuses on SLSA Build L2(+) and L3 is left as an exercise for the
+reader. So what would it take to achieve SLSA Build L3 with GitHub Artifact
+Attestations?
+
+The solution is fairly somewhat so I will just summarize the requirements here
+and leave the details for a later blog post. In order to achieve L3 you, at a
+minimum, would roughly need to do the following things:
+
+- The artifact build would need to be done in a separate job from the call to
+  `attest-build-provenance`. This ensures that the build and provenance signing
+  happen on [separate virtual
+  machines](https://docs.github.com/en/actions/learn-github-actions/understanding-github-actions#the-components-of-github-actions)
+  and that the artifact build steps will not have access to the signing
+  material. This job should **not** have `attestations: write` permissions.
+- The `attest-build-provenance` action is passed a path to the artifact. Jobs
+  execute on separate VMs so the artifact itself would need to be passed from
+  the build job to the attestation job. This will need to be done by uploading
+  the artifact file as a [workflow
+  artifact](https://docs.github.com/en/actions/using-workflows/storing-workflow-data-as-artifacts)
+  from the first job and downloading it in the second job. However, this
+  artifact could be overwritten by concurrently running build steps. So you
+  will need to take a checksum (sha256 etc.) and pass it from the build job to
+  the provenance signing job using job inputs/outputs, and verify it after
+  download.
+
+## Conclusion
+
+Trust in GitHub’s Artifact Attestations trust really lies in the Sigstore
+certificate and its OID claims. The signed SLSA provenance is really only nice
+to have. GitHub ensures that the JSON printed by the `gh` client’s `--format
+json` option is trustworthy simply by omitting any values from the SLSA
+provenance JSON that are not included in the certificate’s OID claims. This
```

Comment:
sigstore-go verifies the subject only - https://github.com/sigstore/sigstore-go/blob/main/pkg/verify/signature.go#L136-L184

Reply:
The GitHub CLI also provides a `github.com/sigstore/sigstore-go/pkg/verify.PolicyBuilder` https://github.com/cli/cli/blob/ff3a59802dfe0957fba65584551783ece554a03a/pkg/cmd/attestation/verification/sigstore.go#L121

The `PolicyBuilder` has checks on the repo owner and/or repo name depending on the user options given
https://github.com/cli/cli/blob/ff3a59802dfe0957fba65584551783ece554a03a/pkg/cmd/attestation/verify/policy.go#L31-L59

So far this is all I've been able to verify that they do.

Reply:
It looks like the extensions are checked by sigstore-go here
https://github.com/sigstore/sigstore-go/blob/817509999e0de65d795fa099010ea1fbdd7b0828/pkg/verify/certificate_identity.go#L133

Reply:
Yea, that's correct, so this should be sufficient assuming you setup the policy to check those extension values. This would ignore the predicate values.

Reply:
It seems the attestations API does check them and will refuse to persist the attestation in the store if the values don't match. So it's possible the `gh` CLI is relying on this check.

I get this error if I use [my fork](https://github.com/ianlewis/attest-build-provenance/blob/e953d6e919ea108bd6225e3aaf617ec4d68870cb/src/main.ts#L27-L28) and generate a predicate with a different workflow ref.

> Error: Error: Failed to persist attestation: Invalid Argument - values do not match: .github/workflows/artifact-attestations.basic.yml != .github/workflows/artifact-attestations.malicious.yml - https://docs.github.com/rest/repos/repos#create-an-attestation

https://github.com/ianlewis/gha-artifact-attestations-test/actions/runs/9170942023/job/25214201468

Reply:
so if there's ever a bug that allows writing to the attestation store, it'd allow an attacker to put anything in the intoto payload.

Reply:
I'll see if I can test their verification code with the attestation store mocked out to verify my assumptions.